### DISCUSSION TITLE
Wix4 Controlling when a call to an extension happens / Virtual Service Accounts

### DISCUSSION THREAD

**Discussion by Trikkitt** [wixtoolset/issues#8658] [Questions] [answered]
I'm trying to use Wix4 to deploy a Windows service.  I want to make use of the Windows virtual service account feature.  However I'm hitting a significant block.

The following code will install the service, but it fails to start.
`
      <Component>
        <File Source="$(var.QControlServer.TargetDir)QControlServer.exe" />
        <ServiceInstall Id="QControlServerService" Name="QControlServer" DisplayName="QControl Server" Description="QControl server application"
                        Start="auto" Type="ownProcess" ErrorControl="normal" Account="NT Service\QControlServer" Vital="yes" Interactive="no"/>        
        <ServiceControl Id="QControlServerService" Name="QControlServer" Start="install" Stop="both" Remove="uninstall" Wait="yes"/>
      </Component>
`

I did some digging and it looks like the virtual service account doesn't have the Logon As a Service right assigned to it by default.  If I wait for my installer to complete the service installation, then manually grant it the right the service will start.  Great - but I can't find a way to assign those rights at the right time within Wix4.

There is a potential option to use the util extension included within Wix.

`
<util:User UpdateIfExists="yes" CreateUser="no" Name="NT Service\QControlServer" LogonAsService="yes"  />
`

Unfortunately I can't figure out how to specify to run this only after the ServiceInstall has completed. It would need to run after the service is installed and before the service start is attempted.  From looking at the execution sequence it appears that these tasks happen one after the other, which is fine.  But I simply can't figure out how to drop the util:User extension action inbetween.  The <InstallExecuteSequence> block appears to allow me to specify that something should happen before or after a given activity, but then it seems like that can only link to a custom action, which in turn appears to only be an external executable rather than the built-in Wix extension.  I really don't want to have to hand craft and deploy additional executables to complete these actions, it just seems excessive given that it should be possible within Wix's capabilities.

Does anyone have any pointers as to how I can make this flow work?

**Comment by bevanweiss** [comment, marked as answer]
The `util:User` functionality is broken up into a few different Custom Actions.  So you'd need to control the scheduling of the appropriate one of those.
https://github.com/wixtoolset/wix/blob/ce73352b1fa1d4f9cded10a0ee410f2e786bd326/src/ext/Util/wixlib/UtilExtension_Platform.wxi#L136-L145

What normally happens is that ConfigureUsers (the Custom Action) is introduced into the execution schedule, it runs and looks through the .MSI tables for user things, and based on desired state and current state it determines which other Custom Actions should be called.

If you look in Orca, you could try to schedule ConfigureUsers between the InstallServices, and the StartServices.
Easiest would be to use Orca to just hand edit first.  Then see if that does the job.  If it does, then you could do the WiX authoring to align.

**Reply by Trikkitt** [reply]
Thanks.  This gave me enough clues to figure it out after a bit of poking around in Orca to establish that the prefix and suffixes were for the custom action.  I ended up adding this to my installer:


`    <InstallExecuteSequence>
      <Custom Action="override Wix4ConfigureUsers_X86" After="InstallServices" />
    </InstallExecuteSequence>
`

Once that was in there along with the component section looking like this:
`      <Component>
        <File Source="$(var.QControlServer.TargetDir)QControlServer.exe" />
        <ServiceInstall Id="QControlServerService" Name="QControlServer" DisplayName="QControl Server" Description="QControl server application"
                        Start="auto" Type="ownProcess" ErrorControl="normal" Account="NT Service\QControlServer" Vital="yes" Interactive="no"/>
        <util:User UpdateIfExists="yes" CreateUser="no" Name="NT Service\QControlServer" LogonAsService="yes"  />
        <ServiceControl Id="QControlServerService" Name="QControlServer" Start="install" Stop="both" Remove="uninstall" Wait="yes"/>
      </Component>
`

My service installed and started successfully using a virtual service account.  Far better than the local service or local system options which are shared between any other installed services and could be over-privileged.

Wix is so powerful, it just feels like some really clever bits aren't documented anywhere, such as overriding the execution order of built-in extensions.  Thanks for your help - I don't think I'd have got there without you!

**Reply by bevanweiss** [reply]
In many areas, WiX is a relatively thin 'shim' to the Windows Installer system.
The issue around documentation is a common open source issue.  Engineers don't tend to like writing documentation, and if they're not being paid, they don't like doing things that they don't like... so creation of documentation falls aside.  Although for it's part, WiX does have better documentation than many other open source projects, which is quite amazing, because it also seems that WiX has a much lower external contribution ratio than other projects I've been a part of...

As an open source project, I'm sure WiX would happily accept any improvements to the documentation.
Whilst Rob has mentioned the override keyword in his Deployment Dojo series, I don't recall seeing it called out in many locations within the WiX schema documentation.
